### PR TITLE
Fix parallelly generating index failure with Mysql

### LINKED ISSUE #22109: Flaky test: `TestRepoCommitsStatusParallel`
### Description

https://drone.gitea.io/go-gitea/gitea/64361/2/14

```log
testlogger.go:77: 2022/12/12 08:53:27 ...pi/v1/repo/status.go:65:NewCommitStatus() [E] [6396ec07-46] CreateCommitStatus: NewCommitStatus[repo_id: 1, user_id: 2, sha: 65f1bf27bc3bf70f64657658635e66094edbcb4d]: insert CommitStatus[/tmp/tmp.nh2kDx6pG3/tests/integration/gitea-integration-mysql/gitea-repositories/user2/repo1.git, 65f1bf27bc3bf70f64657658635e66094edbcb4d]: Error 1062: Duplicate entry '4-1-65f1bf27bc3bf70f64657658635e66094edbcb4d' for key 'UQE_commit_status_repo_sha_index'
--- FAIL: TestRepoCommitsStatusParallel/ParallelCreateStatus_6 (0.38s)
        pull_status_test.go:104: 
            	Error Trace:	/drone/src/integration_test.go:336
            	            				/drone/src/integration_test.go:165
            	            				/drone/src/pull_status_test.go:104
            	            				/drone/src/repo_commits_test.go:141
            	Error:      	Not equal: 
            	            	expected: 201
            	            	actual  : 500
            	Test:       	TestRepoCommitsStatusParallel/ParallelCreateStatus_6
            	Messages:   	Request: POST /api/v1/repos/user2/repo1/statuses/65f1bf27bc3bf70f64657658635e66094edbcb4d?token=95f5e0e052db6781e31573a683234ad48417a1ea
        pull_status_test.go:104: Response: {"message":"","url":"http://localhost:3001/api/swagger"}
```

### Gitea Version

6398ca745aabdc40f10465e71da99c5a3866990e

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

localhost

### Database

Mysql

### PR DESCRIPTION
**Pull request by lunny**
- Fix possible parallel creating commit status index problem and creating issues/pull request index problem work with Mysql5/Mysql8
- Add parallel tests
- Reenable TestRepoCommitsStatusParallel on CI

Fix #22109

### CODE REVIEW

**Comment by a1012112796** [comment]
wonder why not directly using ``SELECT MAX(indx) FROM commit_status WHERE ...`` ?

**Comment by lunny** [comment]
> wonder why not directly using `SELECT MAX(indx) FROM commit_status WHERE ...` ?

That may result in two commit_status having the same index.

File: models/db/index.go [hunk 1]
```diff
@@ -71,10 +71,35 @@ func postgresGetNextResourceIndex(ctx context.Context, tableName string, groupID
...
+	// start a transaction to lock the mysql thread so that LAST_INSERT_ID return the correct value
+	return idx, WithTx(ctx, func(ctx context.Context) error {
+		if _, err := GetEngine(ctx).Exec(fmt.Sprintf("INSERT INTO %s (group_id, max_index) "+
+			"VALUES (?,LAST_INSERT_ID(1)) ON DUPLICATE KEY UPDATE max_index = LAST_INSERT_ID(max_index+1)",
+			// if LAST_INSERT_ID(expr) is used, next time when use LAST_INSERT_ID(), it will be given the expr value
+			tableName), groupID); err != nil {
+			return err
+		}
+
+		_, err := GetEngine(ctx).SQL("SELECT LAST_INSERT_ID()").Get(&idx)
```
Comment (wxiaoguang) [inline models/db/index.go]: IMO:

* no need to use `WithTx` (GetNextResourceIndex is already in transaction)
    * `// GetNextResourceIndex generates a resource index, it must run in the same transaction where the resource is created`
* no need to use `LAST_INSERT_ID`
    * it is not for this case, a simple `SELECT max_index WHERE group_id ... ` should work

File: models/git/commit_status.go [hunk 2]
```diff
@@ -64,48 +64,90 @@ func postgresGetCommitStatusIndex(ctx context.Context, repoID int64, sha string)
...
 	}
 	affected, err := res.RowsAffected()
 	if err != nil {
 		return 0, err
 	}
 	if affected == 0 {
 		// this slow path is only for the first time of creating a resource index
-		_, errIns := e.Exec("INSERT INTO `commit_status_index` (repo_id, sha, max_index) VALUES (?, ?, 0)", repoID, sha)
+		var errIns error
+		for i := 0; i < 5; i++ {
```
Comment (wxiaoguang) [inline models/git/commit_status.go]: Although I wrote the demo code for this `for retry`, actually this retry is wrong for this case (The right deadlock retry should be put out of the transactional function call -- with a new transaction)

The comment could be kept while the code should be reverted.

**Review by wxiaoguang — APPROVED** [review]

**Review by wolfogre — APPROVED** [review]

**Pull request merged**